File: backend-rust/src/middleware/permissions.rs
// ============================================================================
// CYBERSHEPPARD (MicroSIEM) - Permission Helpers
// ============================================================================

use axum::{
    async_trait,
    extract::FromRequestParts,
    http::{request::Parts, StatusCode},
    Json,
};
use serde_json::json;
use crate::middleware::auth::AuthUser;

/// User roles in the system
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum UserRole {
    Admin,
    TeamLeader,
    User,
}

impl UserRole {
    pub fn from_str(role: &str) -> Option<Self> {
        match role.to_lowercase().as_str() {
            "admin" => Some(UserRole::Admin),
            "teamleader" | "team_leader" => Some(UserRole::TeamLeader),
            "user" => Some(UserRole::User),
            _ => None,
        }
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            UserRole::Admin => "admin",
            UserRole::TeamLeader => "teamLeader",
            UserRole::User => "user",
        }
    }
}

/// Permission checker
pub struct Permissions;

impl Permissions {
    /// Check if user is admin
    pub fn is_admin(user: &AuthUser) -> bool {
        user.role == "admin"
    }

    /// Check if user is admin or team leader
    pub fn is_admin_or_team_leader(user: &AuthUser) -> bool {
        user.role == "admin" || user.role == "teamLeader"
    }

    /// Check if user can manage users
    pub fn can_manage_users(user: &AuthUser) -> bool {
        // Admin can manage all users
        // TeamLeader can manage team members (checked at query level)
        user.role == "admin" || user.role == "teamLeader"
    }

    /// Check if user can manage system settings
    pub fn can_manage_system_settings(user: &AuthUser) -> bool {
        user.role == "admin"
    }

    /// Check if user can install/uninstall plugins
    pub fn can_manage_plugins(user: &AuthUser) -> bool {
        user.role == "admin"
    }

    /// Check if user can configure plugins
    pub fn can_configure_plugins(user: &AuthUser) -> bool {
        user.role == "admin" || user.role == "teamLeader"
    }

    /// Check if user can view plugins
    pub fn can_view_plugins(user: &AuthUser) -> bool {
        true // All authenticated users
    }

    /// Check if user can create/edit/delete resources
    pub fn can_manage_resources(user: &AuthUser) -> bool {
        user.role == "admin" || user.role == "teamLeader"
    }

    /// Check if user can execute scans
    pub fn can_execute_scans(user: &AuthUser) -> bool {
        true // All authenticated users
    }

    /// Check if user can view audit logs
    pub fn can_view_audit_logs(user: &AuthUser) -> bool {
        user.role == "admin" || user.role == "teamLeader"
    }

    /// Check if user can manage integrations
    pub fn can_manage_integrations(user: &AuthUser) -> bool {
        user.role == "admin" || user.role == "teamLeader"
    }

    /// Check if user can generate API keys
    pub fn can_generate_api_keys(user: &AuthUser) -> bool {
        true // All users can generate their own API keys
    }

    /// Check if user can view all API keys
    pub fn can_view_all_api_keys(user: &AuthUser) -> bool {
        user.role == "admin"
    }

    /// Check if user can view team API keys
    pub fn can_view_team_api_keys(user: &AuthUser) -> bool {
        user.role == "admin" || user.role == "teamLeader"
    }
}

// ============================================================================
// Role-gated extractors
// ============================================================================
// Questi estrattori applicano l'autorizzazione in modo DICHIARATIVO: basta
// usarli nella firma di un handler (es. `AdminUser(user): AdminUser`) perché
// la route richieda quel ruolo. Se il ruolo non è sufficiente, l'handler non
// viene mai eseguito e si risponde 403. Riutilizzano l'estrazione di AuthUser
// (popolato da auth_middleware), quindi l'assenza di autenticazione dà 401.

fn forbidden(msg: &str) -> (StatusCode, Json<serde_json::Value>) {
    (StatusCode::FORBIDDEN, Json(json!({ "error": msg })))
}

/// Estrattore che richiede ruolo `admin`.
pub struct AdminUser(pub AuthUser);

#[async_trait]
impl<S> FromRequestParts<S> for AdminUser
where
    S: Send + Sync,
{
    type Rejection = (StatusCode, Json<serde_json::Value>);

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
        let user = AuthUser::from_request_parts(parts, state).await?;
        if Permissions::is_admin(&user) {
            Ok(AdminUser(user))
        } else {
            Err(forbidden("Admin access required"))
        }
    }
}

/// Estrattore che richiede ruolo `admin` oppure `teamLeader`.
pub struct ManagerUser(pub AuthUser);

#[async_trait]
impl<S> FromRequestParts<S> for ManagerUser
where
    S: Send + Sync,
{
    type Rejection = (StatusCode, Json<serde_json::Value>);

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
        let user = AuthUser::from_request_parts(parts, state).await?;
        if Permissions::is_admin_or_team_leader(&user) {
            Ok(ManagerUser(user))
        } else {
            Err(forbidden("Admin or team leader access required"))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_user_role_from_str() {
        assert_eq!(UserRole::from_str("admin"), Some(UserRole::Admin));
        assert_eq!(UserRole::from_str("teamLeader"), Some(UserRole::TeamLeader));
        assert_eq!(UserRole::from_str("team_leader"), Some(UserRole::TeamLeader));
        assert_eq!(UserRole::from_str("user"), Some(UserRole::User));
        assert_eq!(UserRole::from_str("invalid"), None);
    }

    #[test]
    fn test_admin_permissions() {
        let admin = AuthUser {
            user_id: 1,
            username: "admin".to_string(),
            role: "admin".to_string(),
            mcp_key_scope: None,
        };

        assert!(Permissions::is_admin(&admin));
        assert!(Permissions::is_admin_or_team_leader(&admin));
        assert!(Permissions::can_manage_users(&admin));
        assert!(Permissions::can_manage_system_settings(&admin));
        assert!(Permissions::can_manage_plugins(&admin));
    }

    #[test]
    fn test_team_leader_permissions() {
        let team_leader = AuthUser {
            user_id: 2,
            username: "leader".to_string(),
            role: "teamLeader".to_string(),
            mcp_key_scope: None,
        };

        assert!(!Permissions::is_admin(&team_leader));
        assert!(Permissions::is_admin_or_team_leader(&team_leader));
        assert!(Permissions::can_manage_users(&team_leader));
        assert!(!Permissions::can_manage_system_settings(&team_leader));
        assert!(!Permissions::can_manage_plugins(&team_leader));
        assert!(Permissions::can_configure_plugins(&team_leader));
    }

    #[test]
    fn test_user_permissions() {
        let user = AuthUser {
            user_id: 3,
            username: "user".to_string(),
            role: "user".to_string(),
            mcp_key_scope: None,
        };

        assert!(!Permissions::is_admin(&user));
        assert!(!Permissions::is_admin_or_team_leader(&user));
        assert!(!Permissions::can_manage_users(&user));
        assert!(!Permissions::can_manage_system_settings(&user));
        assert!(!Permissions::can_manage_plugins(&user));
        assert!(!Permissions::can_configure_plugins(&user));
        assert!(Permissions::can_view_plugins(&user));
        assert!(Permissions::can_execute_scans(&user));
    }

    // ------------------------------------------------------------------
    // Estrattori role-gated: provano l'autorizzazione effettiva (200 vs 403)
    // ------------------------------------------------------------------

    fn parts_with_role(role: &str) -> axum::http::request::Parts {
        let mut req = axum::http::Request::builder().body(()).unwrap();
        req.extensions_mut().insert(AuthUser {
            user_id: 1,
            username: "u".to_string(),
            role: role.to_string(),
            mcp_key_scope: None,
        });
        req.into_parts().0
    }

    #[tokio::test]
    async fn admin_extractor_allows_admin_only() {
        let mut p = parts_with_role("admin");
        assert!(AdminUser::from_request_parts(&mut p, &()).await.is_ok());

        for role in ["teamLeader", "user"] {
            let mut p = parts_with_role(role);
            assert!(
                AdminUser::from_request_parts(&mut p, &()).await.is_err(),
                "role {role} must NOT pass AdminUser"
            );
        }
    }

    #[tokio::test]
    async fn manager_extractor_allows_admin_and_team_leader() {
        for role in ["admin", "teamLeader"] {
            let mut p = parts_with_role(role);
            assert!(
                ManagerUser::from_request_parts(&mut p, &()).await.is_ok(),
                "role {role} must pass ManagerUser"
            );
        }
        let mut p = parts_with_role("user");
        assert!(ManagerUser::from_request_parts(&mut p, &()).await.is_err());
    }

    #[tokio::test]
    async fn extractor_rejects_unauthenticated() {
        // Parts senza AuthUser nelle extensions => 401 (via AuthUser).
        let mut parts = axum::http::Request::builder().body(()).unwrap().into_parts().0;
        assert!(AdminUser::from_request_parts(&mut parts, &()).await.is_err());
    }
}

File: frontend-react/src/utils/permissions.ts
// ============================================================================
// CYBERSHEPPARD (MicroSIEM) - Permission Utilities
// ============================================================================

/**
 * User roles in the system
 * - admin: Full system access
 * - teamLeader: Team management and resource access
 * - user: Basic user access
 */
export type UserRole = 'admin' | 'teamLeader' | 'user';

/**
 * User interface matching backend AuthUser
 */
export interface User {
  id: number;
  username: string;
  email: string;
  role: UserRole;
  team_id?: number;
  managed_by?: number;
  is_active: boolean;
  created_at: string;
}

/**
 * Permission checker class
 */
export class Permissions {
  /**
   * Check if user is admin
   */
  static isAdmin(role: UserRole): boolean {
    return role === 'admin';
  }

  /**
   * Check if user is admin or team leader
   */
  static isAdminOrTeamLeader(role: UserRole): boolean {
    return role === 'admin' || role === 'teamLeader';
  }

  /**
   * Check if user can manage users
   * Admin: All users
   * TeamLeader: Team members only
   */
  static canManageUsers(role: UserRole): boolean {
    return role === 'admin' || role === 'teamLeader';
  }

  /**
   * Check if user can manage system settings
   * Admin only
   */
  static canManageSystemSettings(role: UserRole): boolean {
    return role === 'admin';
  }

  /**
   * Check if user can install/uninstall plugins
   * Admin only
   */
  static canManagePlugins(role: UserRole): boolean {
    return role === 'admin';
  }

  /**
   * Check if user can configure plugins
   * Admin and TeamLeader
   */
  static canConfigurePlugins(role: UserRole): boolean {
    return role === 'admin' || role === 'teamLeader';
  }

  /**
   * Check if user can view plugins
   * All authenticated users
   */
  static canViewPlugins(role: UserRole): boolean {
    return true;
  }

  /**
   * Check if user can create/edit/delete resources (targets, scans, etc.)
   * Admin and TeamLeader
   */
  static canManageResources(role: UserRole): boolean {
    return role === 'admin' || role === 'teamLeader';
  }

  /**
   * Check if user can execute scans
   * All authenticated users
   */
  static canExecuteScans(role: UserRole): boolean {
    return true;
  }

  /**
   * Check if user can view audit logs
   * Admin and TeamLeader
   */
  static canViewAuditLogs(role: UserRole): boolean {
    return role === 'admin' || role === 'teamLeader';
  }

  /**
   * Check if user can manage integrations
   * Admin and TeamLeader
   */
  static canManageIntegrations(role: UserRole): boolean {
    return role === 'admin' || role === 'teamLeader';
  }

  /**
   * Check if user can generate API keys
   * All users (own keys)
   */
  static canGenerateApiKeys(role: UserRole): boolean {
    return true;
  }

  /**
   * Check if user can view all API keys
   * Admin only
   */
  static canViewAllApiKeys(role: UserRole): boolean {
    return role === 'admin';
  }

  /**
   * Check if user can view team API keys
   * Admin and TeamLeader
   */
  static canViewTeamApiKeys(role: UserRole): boolean {
    return role === 'admin' || role === 'teamLeader';
  }

  /**
   * Check if user can edit specific resource
   * Admin: all resources
   * TeamLeader: team resources only
   * User: own resources only
   */
  static canEditResource(
    userRole: UserRole,
    userId: number,
    resourceOwnerId: number,
    resourceTeamId?: number,
    userTeamId?: number
  ): boolean {
    // Admin can edit everything
    if (userRole === 'admin') return true;

    // User owns the resource
    if (userId === resourceOwnerId) return true;

    // TeamLeader can edit team resources
    if (
      userRole === 'teamLeader' &&
      resourceTeamId &&
      userTeamId &&
      resourceTeamId === userTeamId
    ) {
      return true;
    }

    return false;
  }

  /**
   * Get role display name
   */
  static getRoleDisplayName(role: UserRole): string {
    const roleNames: Record<UserRole, string> = {
      admin: 'Administrator',
      teamLeader: 'Team Leader',
      user: 'User',
    };
    return roleNames[role];
  }

  /**
   * Get role badge color for UI
   */
  static getRoleBadgeColor(role: UserRole): string {
    const colors: Record<UserRole, string> = {
      admin: 'red',
      teamLeader: 'blue',
      user: 'gray',
    };
    return colors[role];
  }
}

/**
 * React hook for checking permissions
 */
export function usePermissions(role: UserRole) {
  return {
    isAdmin: Permissions.isAdmin(role),
    isAdminOrTeamLeader: Permissions.isAdminOrTeamLeader(role),
    canManageUsers: Permissions.canManageUsers(role),
    canManageSystemSettings: Permissions.canManageSystemSettings(role),
    canManagePlugins: Permissions.canManagePlugins(role),
    canConfigurePlugins: Permissions.canConfigurePlugins(role),
    canViewPlugins: Permissions.canViewPlugins(role),
    canManageResources: Permissions.canManageResources(role),
    canExecuteScans: Permissions.canExecuteScans(role),
    canViewAuditLogs: Permissions.canViewAuditLogs(role),
    canManageIntegrations: Permissions.canManageIntegrations(role),
    canGenerateApiKeys: Permissions.canGenerateApiKeys(role),
    canViewAllApiKeys: Permissions.canViewAllApiKeys(role),
    canViewTeamApiKeys: Permissions.canViewTeamApiKeys(role),
  };
}

/**
 * Higher-order component for protected routes
 */
export function withPermission<P extends object>(
  Component: React.ComponentType<P>,
  requiredPermission: (role: UserRole) => boolean
) {
  return function ProtectedComponent(props: P & { userRole: UserRole }) {
    const { userRole, ...rest } = props;

    if (!requiredPermission(userRole)) {
      return (
        <div className="flex items-center justify-center h-screen">
          <div className="text-center">
            <h1 className="text-2xl font-bold text-red-600 mb-2">
              Access Denied
            </h1>
            <p className="text-gray-600">
              You don't have permission to access this page.
            </p>
          </div>
        </div>
      );
    }

    return <Component {...(rest as P)} />;
  };
}
